python/neutron-understack/neutron_understack/tests/conftest.py

```diff
@@ -1,26 +1,158 @@
+    return ml2_plugin._make_network_dict(Network(), process_extensions=False)
+
+
+@pytest.fixture
+def network_context(ml2_plugin, network_dict) -> NetworkContext:
+    return NetworkContext(
+        ml2_plugin,
+        "plugin_context",
+        network_dict,
+        segments=[],
+    )
+
+
+@pytest.fixture
+def subnet(request) -> Subnet:
+    req = getattr(request, "param", {})
+    return Subnet(
+        id="aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
+        network_id="bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb",
+        cidr="1.0.0.0/24",
+        external=req.get("external") or False,
+    )
 
 
 @pytest.fixture
-def device_id() -> uuid.UUID:
-    return uuid.uuid4()
+def subnet_dict(ml2_plugin, subnet) -> dict:
+    return ml2_plugin.construct_subnet_dict(subnet)
 
 
 @pytest.fixture
-def mac_address() -> str:
-    return (
-        f"{random.randint(0, 255):02x}:"
-        f"{random.randint(0, 255):02x}:"
-        f"{random.randint(0, 255):02x}:"
-        f"{random.randint(0, 255):02x}:"
-        f"{random.randint(0, 255):02x}:"
-        f"{random.randint(0, 255):02x}"
+def subnet_context(ml2_plugin, subnet_dict) -> SubnetContext:
+    return SubnetContext(
+        ml2_plugin,
+        "plugin_context",
+        subnet_dict,
+        None,
     )
 
 
 @pytest.fixture
-def network_id() -> uuid.UUID:
-    return uuid.uuid4()
+def binding_profile(request) -> str:
+    req = getattr(request, "param", {})
+    return json.dumps(
+        {
+            "local_link_information": [
+                {
+                    "port_id": req.get("port_id")
+                    or "03921f8d-b4de-412e-a733-f8eade4c6268",
+                    "switch_id": "11:22:33:44:55:66",
+                }
+            ]
+        }
+    )
+
+
+@pytest.fixture
+def trunk(subport) -> Trunk:
+    return Trunk(sub_ports=[subport])
+
+
+@pytest.fixture
+def port_binding(binding_profile) -> PortBinding:
+    binding = PortBinding(
+        profile=binding_profile,
+        port=Port(),
+        vif_type=portbindings.VIF_TYPE_OTHER,
+        vnic_type=portbindings.VNIC_BAREMETAL,
+    )
+    return binding
+
+
+@pytest.fixture
+def port_dict(request, port_binding, trunk, ml2_plugin) -> dict:
+    req = getattr(request, "param", {})
+    port_binding.vif_type = req.get("vif_type") or port_binding.vif_type
+    portdict = ml2_plugin.construct_port_dict(port_binding.port)
+    if req.get("trunk"):
+        trunk.port = port_binding.port
+        return extend_port_dict_with_trunk(portdict, port_binding.port)
+    return portdict
+
+
+@pytest.fixture
+def port_context(network_context, port_dict, port_binding, ml2_plugin) -> PortContext:
+    return PortContext(
+        ml2_plugin,
+        "plugin_context",
+        port_dict,
+        network_context,
+        port_binding,
+        None,
+        original_port=copy.deepcopy(port_dict),
+    )
+
+
+@pytest.fixture
+def nautobot_client() -> Nautobot:
+    return MagicMock(spec_set=Nautobot)
```

Comment:
If you do:

```suggestion
@pytest.fixture
def nautobot_client(mocker) -> Nautobot:
    return mocker.MagicMock(spec_set=Nautobot)
```

then you can remove import from unittest as [the MagicMock](https://pytest-mock.readthedocs.io/en/latest/usage.html) is available on the mocker fixture as a convenience method

Reply:
implemented